002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - Represents a read-only snapshot of a module's exports.
+    
+### Operations
+    
+* [ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)
+    - Creates a SourceTextModuleRecord from source code.
+    
+* [HostResolveImportedModule](https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
```

Comment:
Even if people are not using `babel`, we need concrete examples of what is going to break. The intent is to have well defined semantics of what will break just like Strict Mode. Another intent is to ensure that the upgrade path does not accidentally leak these semantics into older code, just like Strict Mode. We also want to avoid the problem of breaking the ability to use legacy code in a vein similar to Python3. To that end, unless people can clearly form a statement on:

* semantics leaking into CJS
* why ES6 <=> CJS *is* unable to be presented to each other at *runtime*. The interop is not about tooling at compile time.
* why breaking the ability to use legacy code is not going to be like Python3

I will be mostly ignoring this comment. I will remind all that in this PR Node is the first class citizen with existing semantics, breaking those semantics/expectations will need to avoid at least the 3 points mentioned above.